tests/fv/critical_test.go

```diff
@@ -0,0 +1,139 @@
+		k8sEtcd          *containers.Container
+		apiserver        *containers.Container
+		calicoClient     *client.Client
+		k8sClient        *kubernetes.Clientset
+
+		policyName    string
+		genPolicyName string
+	)
+
+	BeforeEach(func() {
+		// Run etcd.
+		calicoEtcd = testutils.RunEtcd()
+		calicoClient = GetCalicoClient(calicoEtcd.IP)
+		err := calicoClient.EnsureInitialized()
+		Expect(err).NotTo(HaveOccurred())
+
+		// Run apiserver.
+		k8sEtcd = testutils.RunEtcd()
+		apiserver = testutils.RunK8sApiserver(k8sEtcd.IP)
+
+		// Write out a kubeconfig file
+		kfconfigfile, err := ioutil.TempFile("", "ginkgo-policycontroller")
+		Expect(err).NotTo(HaveOccurred())
+		defer os.Remove(kfconfigfile.Name())
+		data := fmt.Sprintf(KubeconfigTemplate, apiserver.IP)
+		kfconfigfile.Write([]byte(data))
+
+		policyController = RunPolicyController(calicoEtcd.IP, kfconfigfile.Name())
+
+		k8sClient, err = GetK8sClient(kfconfigfile.Name())
+		Expect(err).NotTo(HaveOccurred())
+
+		// TODO: Use upcoming port checker functions to wait until apiserver is responding to requests.
+		time.Sleep(time.Second * 15)
+
+		// Create a Kubernetes NetworkPolicy.
+		policyName = "jelly"
+		genPolicyName = "knp.default.default." + policyName
+		var np *extensions.NetworkPolicy
+		np = &extensions.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      policyName,
+				Namespace: "default",
+			},
+			Spec: extensions.NetworkPolicySpec{
+				PodSelector: metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"fools": "gold",
+					},
+				},
+			},
+		}
+		err = k8sClient.Extensions().RESTClient().
+			Post().
+			Resource("networkpolicies").
+			Namespace("default").
+			Body(np).
+			Do().Error()
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait for it to appear in Calico's etcd.
+		Eventually(func() *api.Policy {
+			policy, _ := calicoClient.Policies().Get(api.PolicyMetadata{Name: genPolicyName})
+			return policy
+		}).ShouldNot(BeNil())
+	})
+	Context("when apiserver goes down momentarily and data is removed from calico's etcd", func() {
+		It("should eventually add the data to calico's etcd", func() {
+			Stop(apiserver)
+			err := calicoClient.Policies().Delete(api.PolicyMetadata{Name: genPolicyName})
+			Expect(err).ShouldNot(HaveOccurred())
+			Start(apiserver)
+			Eventually(func() error {
+				_, err := calicoClient.Policies().Get(api.PolicyMetadata{Name: genPolicyName})
+				return err
+			}, time.Second*15, 500*time.Millisecond).ShouldNot(HaveOccurred())
+		})
+	})
+	Context("when calico's etcd goes down momentarily and data is removed from k8s-apiserver", func() {
+		It("should eventually remove the data from calico's etcd", func() {
+			// Delete the Policy.
+			Stop(calicoEtcd)
+			err := k8sClient.Extensions().RESTClient().
+				Delete().
+				Resource("networkpolicies").
+				Namespace("default").
+				Name(policyName).
+				Do().Error()
+			Expect(err).NotTo(HaveOccurred())
+
+			Start(calicoEtcd)
```

Comment:
I wonder if there is a race condition here that we need to account for - what if etcd starts up before the policy controller has processed the update and thus we're not actually testing what we expect?

Can probably mitigate this scenario with a quick sleep here, but that feels gross.  Not sure what other options we have though.

Reply:
If it needs to know policy controller internal state to test correctly, perhaps it should be written as a UT as well? I'll add a timeout for now...